Unit tests fails on Python 3.15.0b1

Latest main: 7ce55df0bc965c6605dd89bdb4da91610624290a

See also https://bugzilla.redhat.com/show_bug.cgi?id=2477406 for failure on release 4.15.0
```
jonny in typing_extensions/src on  main [!⇡] via 🐍 v3.14.5
❯ /home/jonny/Downloads/Python-3.15.0b1/python -m unittest discover
..............................................................................................................................................EEEE................................................s....ss....s............................E....F.................................................................................................................................................................EE.......................s.................................s.......................................F...s.s...ss......ssss.....s...............................s........F.....
======================================================================
ERROR: test_annotation_and_optional_default (test_typing_extensions.GetTypeHintTests.test_annotation_and_optional_default) (annotation='typing.Unpack[~Ts]', as_str=True, wrap_optional=False, none_default=False, expected_type_hints={'x': typing.Unpack[~Ts]})
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jonny/projects/typing_extensions/src/test_typing_extensions.py", line 1823, in test_annotation_and_optional_default
    type_hints = get_type_hints(func, globals(), locals(), include_extras=True)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 2503, in get_type_hints
    value = _eval_type(value, globalns, localns, (), format=format, owner=obj, prefer_fwd_module=True)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 475, in _eval_type
    return evaluate_forward_ref(t, globals=globalns, locals=localns,
                                type_params=type_params, owner=owner,
                                _recursive_guard=recursive_guard, format=format)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 1041, in evaluate_forward_ref
    value = forward_ref.evaluate(globals=globals, locals=locals,
                                 type_params=type_params, owner=owner, format=format)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/annotationlib.py", line 207, in evaluate
    return eval(code, globals=globals, locals=locals)
  File "<string>", line 1, in <module>
TypeError: bad operand type for unary ~: 'typing.TypeVarTuple'

======================================================================
ERROR: test_annotation_and_optional_default (test_typing_extensions.GetTypeHintTests.test_annotation_and_optional_default) (annotation='typing.Unpack[~Ts] | None', as_str=True, wrap_optional=True, none_default=False, expected_type_hints={'x': typing.Unpack[~Ts] | None})
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jonny/projects/typing_extensions/src/test_typing_extensions.py", line 1823, in test_annotation_and_optional_default
    type_hints = get_type_hints(func, globals(), locals(), include_extras=True)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 2503, in get_type_hints
    value = _eval_type(value, globalns, localns, (), format=format, owner=obj, prefer_fwd_module=True)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 475, in _eval_type
    return evaluate_forward_ref(t, globals=globalns, locals=localns,
                                type_params=type_params, owner=owner,
                                _recursive_guard=recursive_guard, format=format)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 1041, in evaluate_forward_ref
    value = forward_ref.evaluate(globals=globals, locals=locals,
                                 type_params=type_params, owner=owner, format=format)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/annotationlib.py", line 207, in evaluate
    return eval(code, globals=globals, locals=locals)
  File "<string>", line 1, in <module>
TypeError: bad operand type for unary ~: 'typing.TypeVarTuple'

======================================================================
ERROR: test_annotation_and_optional_default (test_typing_extensions.GetTypeHintTests.test_annotation_and_optional_default) (annotation='typing.Unpack[~Ts]', as_str=True, wrap_optional=False, none_default=True, expected_type_hints={'x': typing.Unpack[~Ts]})
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jonny/projects/typing_extensions/src/test_typing_extensions.py", line 1823, in test_annotation_and_optional_default
    type_hints = get_type_hints(func, globals(), locals(), include_extras=True)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 2503, in get_type_hints
    value = _eval_type(value, globalns, localns, (), format=format, owner=obj, prefer_fwd_module=True)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 475, in _eval_type
    return evaluate_forward_ref(t, globals=globalns, locals=localns,
                                type_params=type_params, owner=owner,
                                _recursive_guard=recursive_guard, format=format)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 1041, in evaluate_forward_ref
    value = forward_ref.evaluate(globals=globals, locals=locals,
                                 type_params=type_params, owner=owner, format=format)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/annotationlib.py", line 207, in evaluate
    return eval(code, globals=globals, locals=locals)
  File "<string>", line 1, in <module>
TypeError: bad operand type for unary ~: 'typing.TypeVarTuple'

======================================================================
ERROR: test_annotation_and_optional_default (test_typing_extensions.GetTypeHintTests.test_annotation_and_optional_default) (annotation='typing.Unpack[~Ts] | None', as_str=True, wrap_optional=True, none_default=True, expected_type_hints={'x': typing.Unpack[~Ts] | None})
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jonny/projects/typing_extensions/src/test_typing_extensions.py", line 1823, in test_annotation_and_optional_default
    type_hints = get_type_hints(func, globals(), locals(), include_extras=True)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 2503, in get_type_hints
    value = _eval_type(value, globalns, localns, (), format=format, owner=obj, prefer_fwd_module=True)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 475, in _eval_type
    return evaluate_forward_ref(t, globals=globalns, locals=localns,
                                type_params=type_params, owner=owner,
                                _recursive_guard=recursive_guard, format=format)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/typing.py", line 1041, in evaluate_forward_ref
    value = forward_ref.evaluate(globals=globals, locals=locals,
                                 type_params=type_params, owner=owner, format=format)
  File "/home/jonny/Downloads/Python-3.15.0b1/Lib/annotationlib.py", line 207, in evaluate
    return eval(code, globals=globals, locals=locals)
  File "<string>", line 1, in <module>
TypeError: bad operand type for unary ~: 'typing.TypeVarTuple'

======================================================================
ERROR: test_constructor (test_typing_extensions.NoExtraItemsTests.test_constructor)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jonny/projects/typing_extensions/src/test_typing_extensions.py", line 7976, in test_constructor
    self.assertIs(self.sentinel_type, type(self.sentinel_type)())
                                      ~~~~~~~~~~~~~~~~~~~~~~~~^^
TypeError: sentinel expected 1 argument, got 0

======================================================================
ERROR: test_sentinel_deprecated (test_typing_extensions.TestSentinels.test_sentinel_deprecated)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jonny/projects/typing_extensions/src/test_typing_extensions.py", line 9649, in test_sentinel_deprecated
    class SentinelSubclass(Sentinel):
        pass
TypeError: type 'sentinel' is not an acceptable base type

======================================================================
ERROR: test_sentinel_deprecated_explicit_repr (test_typing_extensions.TestSentinels.test_sentinel_deprecated_explicit_repr)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jonny/projects/typing_extensions/src/test_typing_extensions.py", line 9604, in test_sentinel_deprecated_explicit_repr
    sentinel_explicit_repr = sentinel('sentinel_explicit_repr', repr='explicit_repr')
TypeError: sentinel() takes no keyword arguments

======================================================================
FAIL: test_repr (test_typing_extensions.NoExtraItemsTests.test_repr)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jonny/projects/typing_extensions/src/test_typing_extensions.py", line 8017, in test_repr
    self.assertEqual(repr(NoExtraItems), f"{mod_name}.NoExtraItems")
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 'NoExtraItems' != 'typing.NoExtraItems'
- NoExtraItems
+ typing.NoExtraItems
? +++++++


======================================================================
FAIL: test_repr (test_typing_extensions.TypeVarTupleTests.test_repr)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jonny/projects/typing_extensions/src/test_typing_extensions.py", line 6802, in test_repr
    self.assertEqual(repr(Ts), 'Ts')
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
AssertionError: '~Ts' != 'Ts'
- ~Ts
? -
+ Ts


======================================================================
FAIL: test_repr (test_typing_extensions.UnpackTests.test_repr)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jonny/projects/typing_extensions/src/test_typing_extensions.py", line 6604, in test_repr
    self.assertEqual(repr(Unpack[Ts]), f'{Unpack.__module__}.Unpack[Ts]')
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 'typing.Unpack[~Ts]' != 'typing.Unpack[Ts]'
- typing.Unpack[~Ts]
?               -
+ typing.Unpack[Ts]


----------------------------------------------------------------------
Ran 571 tests in 0.519s

FAILED (failures=3, errors=7, skipped=16)
```

--- a/src/test_typing_extensions.py
+++ b/src/test_typing_extensions.py
@@ -134,6 +134,10 @@ TYPING_3_13_0_RC = sys.version_info[:4] >= (3, 13, 0, "candidate")
 
 TYPING_3_14_0 = sys.version_info[:3] >= (3, 14, 0)
 
+TYPING_3_15_0 = sys.version_info[:3] >= (3, 15, 0)
+
+TYPING_3_15_0_BETA_1 = sys.version_info[:5] == (3, 15, 0, 'beta', 1)
+
 # https://github.com/python/cpython/pull/27017 was backported into some 3.9 and 3.10
 # versions, but not all
 HAS_FORWARD_MODULE = "module" in inspect.signature(typing._type_check).parameters
@@ -1761,8 +1765,6 @@ class GetTypeHintTests(BaseTestCase):
             Optional[annotation]     : Optional[annotation],
             Union[str, None, str]   : Optional[str],
             Unpack[Tuple[int, None]]: Unpack[Tuple[int, None]],
-            # Note: A starred *Ts will use typing.Unpack in 3.11+ see Issue #485
-            Unpack[Ts]              : Unpack[Ts],
         }
         # contains a ForwardRef, TypeVar(~prefix) or no expression
         do_not_stringify_cases = {
@@ -1779,6 +1781,12 @@ class GetTypeHintTests(BaseTestCase):
             Union["annotation", T_default]     : Union[annotation, T_default],
             Annotated["annotation", "nested"]  : Annotated[Union[int, None], "data", "nested"],
         }
+        # Note: A starred *Ts will use typing.Unpack in 3.11+ see Issue #485
+        if TYPING_3_15_0:
+            # The repr is typing.Unpack[~Ts], which cannot be evaluated.
+            do_not_stringify_cases[Unpack[Ts]] = Unpack[Ts]
+        else:
+            cases[Unpack[Ts]] = Unpack[Ts]
         if TYPING_3_10_0:  # cannot construct UnionTypes before 3.10
             do_not_stringify_cases["str | NoneAlias | StrAlias"] = str | None
             cases[str | None] = Optional[str]
@@ -6599,9 +6607,15 @@ class UnpackTests(BaseTestCase):
         with self.assertRaises(TypeError):
             Unpack()
 
+    @skipIf(TYPING_3_15_0, "repr changed in 3.15")
     def test_repr(self):
         Ts = TypeVarTuple('Ts')
         self.assertEqual(repr(Unpack[Ts]), f'{Unpack.__module__}.Unpack[Ts]')
+
+    @skipUnless(TYPING_3_15_0, "repr changed in 3.15")
+    def test_repr_py315(self):
+        Ts = TypeVarTuple('Ts')
+        self.assertEqual(repr(Unpack[Ts]), f'{Unpack.__module__}.Unpack[~Ts]')
 
     def test_cannot_subclass_vars(self):
         with self.assertRaises(TypeError):
@@ -6797,9 +6811,15 @@ class TypeVarTupleTests(BaseTestCase):
         Ys = TypeVarTuple('Ys')
         self.assertNotEqual(Xs, Ys)
 
+    @skipIf(TYPING_3_15_0, "repr changed in 3.15")
     def test_repr(self):
         Ts = TypeVarTuple('Ts')
         self.assertEqual(repr(Ts), 'Ts')
+
+    @skipUnless(TYPING_3_15_0, "repr changed in 3.15")
+    def test_repr_py315(self):
+        Ts = TypeVarTuple('Ts')
+        self.assertEqual(repr(Ts), '~Ts')
 
     def test_no_redefinition(self):
         self.assertNotEqual(TypeVarTuple('Ts'), TypeVarTuple('Ts'))
@@ -7973,7 +7993,6 @@ class SentinelTestsMixin:
         self.assertIsInstance(self.sentinel_type.__doc__, str)
 
     def test_constructor(self):
-        self.assertIs(self.sentinel_type, type(self.sentinel_type)())
         with self.assertRaises(TypeError):
             type(self.sentinel_type)(1)
 
@@ -8009,12 +8028,17 @@ class NoDefaultTests(SentinelTestsMixin, BaseTestCase):
 class NoExtraItemsTests(SentinelTestsMixin, BaseTestCase):
     sentinel_type = NoExtraItems
 
+    @skipIf(TYPING_3_15_0, "repr changed in 3.15")
     def test_repr(self):
         if hasattr(typing, 'NoExtraItems'):
-            mod_name = 'typing'
+            expected_repr = "typing.NoExtraItems"
         else:
-            mod_name = "typing_extensions"
-        self.assertEqual(repr(NoExtraItems), f"{mod_name}.NoExtraItems")
+            expected_repr = "typing_extensions.NoExtraItems"
+        self.assertEqual(repr(NoExtraItems), expected_repr)
+
+    @skipUnless(TYPING_3_15_0, "repr changed in 3.15")
+    def test_repr_py315(self):
+        self.assertEqual(repr(NoExtraItems), "NoExtraItems")
 
 
 class TypeVarInferVarianceTests(BaseTestCase):
@@ -9537,11 +9561,13 @@ class EvaluateForwardRefTests(BaseTestCase):
             float,
         )
         self.assertIs(evaluate_forward_ref(typing.ForwardRef("int"), globals={"int": str}), str)
+
         import builtins
 
-        from test import support
-        with support.swap_attr(builtins, "int", dict):
-            self.assertIs(evaluate_forward_ref(typing.ForwardRef("int")), dict)
+        old_int = builtins.int
+        self.addCleanup(lambda: setattr(builtins, "int", old_int))
+        builtins.int = dict
+        self.assertIs(evaluate_forward_ref(typing.ForwardRef("int")), dict)
 
     def test_nested_strings(self):
         # This variable must have a different name TypeVar
@@ -9599,10 +9625,9 @@ class TestSentinels(BaseTestCase):
         self.assertEqual(sentinel_no_repr.__name__, 'sentinel_no_repr')
         self.assertEqual(repr(sentinel_no_repr), 'sentinel_no_repr')
 
-    def test_sentinel_deprecated_explicit_repr(self):
-        with self.assertWarnsRegex(DeprecationWarning, r"'repr' parameter is deprecated and will be removed"):
-            sentinel_explicit_repr = sentinel('sentinel_explicit_repr', repr='explicit_repr')
-
+    @skipIf(TYPING_3_15_0_BETA_1, reason="'repr' parameter is not yet available in 3.15.0b1")
+    def test_sentinel_explicit_repr(self):
+        sentinel_explicit_repr = sentinel('sentinel_explicit_repr', repr='explicit_repr')
         self.assertEqual(repr(sentinel_explicit_repr), 'explicit_repr')
 
     @skipIf(sys.version_info < (3, 10), reason='New unions not available in 3.9')
@@ -9644,6 +9669,7 @@ class TestSentinels(BaseTestCase):
             ):
                 self.assertIs(anonymous_sentinel, pickle.loads(pickle.dumps(anonymous_sentinel, protocol=proto)))
 
+    @skipIf(TYPING_3_15_0, reason='Deprecated sentinel APIs were removed in 3.15')
     def test_sentinel_deprecated(self):
         with self.assertWarnsRegex(DeprecationWarning, r"Subclassing sentinel is deprecated"):
             class SentinelSubclass(Sentinel):
@@ -9655,6 +9681,18 @@ class TestSentinels(BaseTestCase):
             my_sentinel = Sentinel(name="my_sentinel")
         with self.assertWarnsRegex(DeprecationWarning, r"Setting attribute 'foo' on sentinel objects is deprecated"):
             my_sentinel.foo = "bar"
+
+    @skipUnless(TYPING_3_15_0, reason='Deprecated sentinel APIs are available before 3.15')
+    def test_sentinel_removed_deprecated_apis(self):
+        with self.assertRaises(TypeError):
+            class SentinelSubclass(Sentinel):
+                pass
+        with self.assertRaises(TypeError):
+            sentinel()
+        with self.assertRaises(TypeError):
+            Sentinel(name="my_sentinel")
+        with self.assertRaises(AttributeError):
+            sentinel('my_sentinel').foo = "bar"
 
 
 def load_tests(loader, tests, pattern):

--- a/src/test_typing_extensions.py
+++ b/src/test_typing_extensions.py
@@ -9560,10 +9560,8 @@ class EvaluateForwardRefTests(BaseTestCase):
 
         import builtins
 
-        old_int = builtins.int
-        self.addCleanup(lambda: setattr(builtins, "int", old_int))
-        builtins.int = dict
-        self.assertIs(evaluate_forward_ref(typing.ForwardRef("int")), dict)
+        with patch.object(builtins, "int", dict):
+             self.assertIs(evaluate_forward_ref(typing.ForwardRef("int")), dict)
 
     def test_nested_strings(self):
         # This variable must have a different name TypeVar

--- a/src/test_typing_extensions.py
+++ b/src/test_typing_extensions.py
@@ -8030,11 +8030,7 @@ class NoExtraItemsTests(SentinelTestsMixin, BaseTestCase):
 
     @skipIf(TYPING_3_15_0, "repr changed in 3.15")
     def test_repr(self):
-        if hasattr(typing, 'NoExtraItems'):
-            expected_repr = "typing.NoExtraItems"
-        else:
-            expected_repr = "typing_extensions.NoExtraItems"
-        self.assertEqual(repr(NoExtraItems), expected_repr)
+        self.assertEqual(repr(NoExtraItems), "typing_extensions.NoExtraItems")
 
     @skipUnless(TYPING_3_15_0, "repr changed in 3.15")
     def test_repr_py315(self):

--- a/src/typing_extensions.py
+++ b/src/typing_extensions.py
@@ -204,13 +204,6 @@ else:
                 __name = name
             if __name is _sentinel_placeholder:
                 raise TypeError("First parameter 'name' is required")
-            if repr is not None:
-                warnings.warn(
-                    "The 'repr' parameter is deprecated "
-                    "and will be removed in Python 3.15.",
-                    DeprecationWarning,
-                    stacklevel=2,
-                )
 
             self.__name__ = __name
             self._repr = repr if repr is not None else __name

